File: leetcode-04/solutions/cpp/0923.cpp
class Solution {
 public:
  int threeSumMulti(vector<int>& A, int target) {
    long ans = 0;

    unordered_map<int, long> map;
    for (int a : A) map[a]++;

    for (auto& [i, x] : map)
      for (auto& [j, y] : map) {
        int k = target - i - j;
        if (!map.count(k)) continue;
        if (i == j && j == k)
          ans += x * (x - 1) * (x - 2) / 6;
        else if (i == j && j != k)
          ans += x * (x - 1) / 2 * map[k];
        else if (i < j && j < k)
          ans += x * y * map[k];
      }

    return ans % int(1e9 + 7);
  }
};

File: leetcode-04/solutions/python3/0923.py
class Solution:
    def threeSumMulti(self, A: List[int], target: int) -> int:
        ans = 0

        dict = collections.Counter(A)

        for i, x in dict.items():
            for j, y in dict.items():
                k = target - i - j
                if k not in dict:
                    continue
                if i == j and j == k:
                    ans += x * (x - 1) * (x - 2) // 6
                elif i == j and j != k:
                    ans += x * (x - 1) // 2 * dict[k]
                elif i < j and j < k:
                    ans += x * y * dict[k]

        return ans % int(1e9 + 7)
